ledger/complete/wal/checkpoint_v6_reader.go

```diff
@@ -0,0 +1,340 @@
+package wal
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"path"
+
+	"github.com/onflow/flow-go/ledger/complete/mtrie/flattener"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/node"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+)
+
+// ReadCheckpointV6 reads checkpoint file from a main file and 17 file parts.
+// the main file stores:
+// 		1. version
+//		2. checksum of each part file (17 in total)
+// 		3. checksum of the main file itself
+// 	the first 16 files parts contain the trie nodes below the subtrieLevel
+//	the last part file contains the top level trie nodes above the subtrieLevel and all the trie root nodes.
+func ReadCheckpointV6(dir string, fileName string) ([]*trie.MTrie, error) {
+	// TODO: read the main file and check the version
+	// TODO: read the checksum of 17 file parts
+
+	headerPath := filePathHeader(dir, fileName)
+	subtrieChecksums, topTrieChecksum, err := readHeader(headerPath)
+	if err != nil {
+		return nil, fmt.Errorf("could not read header: %w", err)
+	}
+
+	subtrieNodes, err := readSubTriesConcurrently(dir, fileName, subtrieChecksums)
+	if err != nil {
+		return nil, fmt.Errorf("could not read subtrie from dir: %w", err)
+	}
+
+	// validate the subtrie node checksum
+	tries, err := readTopLevelTries(dir, fileName, subtrieNodes, topTrieChecksum)
+	if err != nil {
+		return nil, fmt.Errorf("could not read top level nodes or tries: %w", err)
+	}
+
+	return tries, nil
+}
+
+func filePathHeader(dir string, fileName string) string {
+	return path.Join(dir, fileName)
+}
+
+func filePathSubTries(dir string, fileName string, index int) (string, string, error) {
+	if index < 0 || index > 15 {
+		return "", "", fmt.Errorf("index must be between 1 to 16, but got %v", index)
+	}
```

Comment:
I think it's good to keep this check inside this function in case the caller forgot to call it, as this function is being used in multiple places, such as [the check here ](https://github.com/onflow/flow-go/pull/3321/files#diff-2ff6507ee260a1e37156b4cae4cc259e7688bbd0007993b20fa193824073105aR102)and[ the test case here](https://github.com/onflow/flow-go/pull/3321/files#diff-7c9a29fd60f7021a8fd387c9e44b5c42dbeac85f35e4466de4b413208ba0444dR169).

I will also change this function to take additional `subtrieLevel` parameter, so that we can replace the hardcoded `15` here to be more flexible.